### PR TITLE
SOL-150689: Fix flaky e2e-monitoring F4 txMsgRate assertion (get-message-rates)

### PR DESCRIPTION
**Pull request by helen-zhu-solace**
## Summary

Fix flaky `e2e-monitoring` F4 `txMsgRate` assertion in `get-message-rates`. Splits rx/tx thresholds and polls 5× for peak instead of asserting a single SEMP sample.

Fixes [SOL-150715](https://sol-jira.atlassian.net/browse/SOL-150715).

## Type of Change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Documentation update
- [ ] Refactoring (no functional changes)
- [ ] Performance improvement
- [ ] Dependency update

## Motivation and Context

The F4 sustained-traffic check `txMsgRate ≥ 80` flakes on unrelated branches (`wmaryam/SOL-148434`, `hzhu/sol-150248`). Observed `txMsgRate` values: **57, 68, 78, 88** — straddling the 80 line. `rxMsgRate` (~1150) has huge headroom and never flakes.

Root cause: `get-message-rates` returns a single short-window SEMP-averaged sample. `txMsgRate` (delivery to the F3 receiver) is inherently lower and noisier than the publish-side aggregate `rxMsgRate`, and was not empirically characterized in SOL-150024 — only `rxMsgRate` was.

## Changes Made

- Split F4 thresholds: `F4_RX_THRESHOLD=80` (unchanged behaviour), `F4_TX_THRESHOLD=40` (covers observed 57 floor with margin).
- Replace single-read assertion with **peak-of-5 polling** (5 samples × 1 s ≈ 5 s window) in both `test/e2e-monitoring/tool-tests.sh` and `test/e2e-monitoring/verify-fixtures.sh`.
- Log line now includes all samples and computed peaks for diagnosability.
- README updated: F4 fixtures-table row and "F4 — message rates" section now document the empirical 57–88 tx band, the lower tx threshold, and the polling rationale (cites SOL-150715).

## Testing

**Test Environment:**
- Go version: N/A (bash test-suite change only)
- OS: Linux
- Broker version (if applicable): same as e2e-monitoring suite

**Tests Performed:**
- [ ] Unit tests added/updated
- [ ] Integration tests added/updated (if applicable)
- [x] E2E tests added/updated (if applicable)
- [ ] Tested locally with `go test -race ./...`
- [x] Tested with real Solace broker

**Test Coverage:**
- `bash -n` syntax check on both modified shell files.
- Local e2e-monitoring suite run against the two test brokers — F4 assertions pass on both `broker-a` and `broker-b`; log line shows the 5 samples + peaks as expected.
- Flake characterization in the ticket already covers the failure mode the new thresholds + polling are designed to absorb (57–88 tx range).

## Breaking Changes

N/A.

### Code Quality
- [x] Code follows the coding standards
- [x] Self-review completed
- [x] Code is well-commented (explains "why", not "what")
- [x] Complex logic has explanatory comments
- [x] No commented-out code or debug statements left in

### Security
- [x] No credentials in code
- [x] Followed secure logging rules (no new logging of sensitive fields; existing `log_info` only)
- [x] Credential-carrying types implement `slog.LogValuer` (N/A — bash only)
- [x] No security vulnerabilities introduced

### Testing
- [x] All tests pass locally
- [x] No race conditions detected
- [x] Edge cases covered by tests
- [x] Error paths tested (MCP call failure still aborts the test)
- [x] Test names clearly describe what is being tested

### Documentation
- [x] README.md updated (F4 row + timing section)
- [ ] docs/ updated (no architecture/design changes)
- [ ] godoc comments added/updated (N/A — bash)
- [ ] CHANGELOG.md updated (no CHANGELOG in this repo)
- [ ] Examples updated (N/A)

### Git & CI
- [ ] All commits are signed off (DCO: `git commit -s`)
- [x] Commits have clear, descriptive messages
- [x] Branch is up to date with main
- [x] No merge conflicts
- [x] CI checks passing (pending push)

### Breaking Changes (if applicable)
- N/A

## Screenshots (if applicable)

N/A — log line example:

```
[INFO] get-message-rates [broker-b]: samples=[rx=1148,tx=68 rx=1162,tx=74 rx=1155,tx=82 rx=1170,tx=79 rx=1158,tx=71] peakRx=1170 peakTx=82 (rx≥80, tx≥40)
```

## Additional Notes

The 25 s F4 settle window is unchanged — the new ~5 s polling sits *on top of* settle, not in place of it. Total F4 timing budget grows by ~5 s per broker per file (≈ 20 s across `tool-tests.sh` + `verify-fixtures.sh` × 2 brokers), which is well within CI budget.

[SOL-150715]: https://sol-jira.atlassian.net/browse/SOL-150715?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by solace-aross — COMMENTED** [review]
Solid fix, and the README rationale with the empirical 57-88 band is exactly the documentation this kind of threshold change needs. Two questions inline, neither blocking. Approving is pending the e2e-monitoring job going green since that run is the real proof here.

File: test/e2e-monitoring/tool-tests.sh [hunk 1]
```diff
@@ -293,28 +293,44 @@ test_list_client_subscriptions_pagination_a() { test_list_client_subscriptions_p
...
-    log_info "get-message-rates [$broker]: rxMsgRate=$rx txMsgRate=$tx (threshold ≥ $F4_RATE_THRESHOLD)"
-    assert_json_field "$content" ".rates.data.rxMsgRate >= $F4_RATE_THRESHOLD" "true" \
-        "get-message-rates [$broker]: rxMsgRate must be ≥ $F4_RATE_THRESHOLD (got $rx)" || return 1
-    assert_json_field "$content" ".rates.data.txMsgRate >= $F4_RATE_THRESHOLD" "true" \
-        "get-message-rates [$broker]: txMsgRate must be ≥ $F4_RATE_THRESHOLD (got $tx)" || return 1
+    local response content rx tx peak_rx=0 peak_tx=0 i
+    local samples=()
+    for ((i = 1; i <= F4_SAMPLE_COUNT; i++)); do
+        response=$(mcp_call_tool "get-message-rates" \
+            "$(jq -nc --arg b "$broker" '{broker:$b,msgVpnName:"default"}')") || return 1
```
Comment (solace-aross) [inline test/e2e-monitoring/tool-tests.sh:318]: The loop makes 5 calls where we previously made 1, and any single failed call still aborts the test here. That multiplies exposure to transient transport failures by five, which could trade one flake source for another. Did you consider tolerating individual sample failures as long as at least one sample lands? Fine as-is if MCP/SEMP calls have been reliable in CI (the flaky part was the rate values, not the calls). Same applies to the loop in verify-fixtures.sh.

File: test/e2e-monitoring/tool-tests.sh [hunk 2]
```diff
@@ -293,28 +293,44 @@ test_list_client_subscriptions_pagination_a() { test_list_client_subscriptions_p
...
+# 5 times over ~5 s and assert the peak ≥ a lower, empirically-grounded floor.
+# F4 is settled by the orchestrator's verify-fixtures step (which waits ≥ 25 s
+# on F4_READY_EPOCH) before this file runs, and the F4 driver keeps publishing
+# throughout — so the rates are live when read here. No pagination or
+# VPN-scoping variants (VPN-level tool).
 # Envelope: {"rates":{"data":{...}}} — a single object.
 
-F4_RATE_THRESHOLD=80
+F4_RX_THRESHOLD=80
+F4_TX_THRESHOLD=40
```
Comment (solace-aross) [inline test/e2e-monitoring/tool-tests.sh]: Peak-of-5 alone lifts the asserted statistic well above the observed 57 floor (the peak of five draws from a 57-88 band will almost always clear 70), and the threshold also dropped to 40. Stacked, a sustained drop to ~45 tx still passes. If the intent is liveness-only (dead F3 receiver reads ~0, still caught), that is fine. If we want the assertion to keep any regression signal, one of the two margins would do it, e.g. peak >= 50. Your call.

**Review by solace-aross — DISMISSED** [review]
All checks green, including the e2e-monitoring run that exercises this change. My two inline questions are non-blocking; fine to address in a follow-up or resolve with rationale.

File: test/e2e-monitoring/tool-tests.sh (hunk 1 above)
Comment (helen-zhu-solace) [reply test/e2e-monitoring/tool-tests.sh:318]: Leaving as-is for now as MCP/SEMP calls haven't been a source of CI flakes in this suite (the historical failures were all rate-value based, never call errors). If MCP-call flakes show up later, will revisit

File: test/e2e-monitoring/tool-tests.sh (hunk 2 above)
Comment (helen-zhu-solace) [reply test/e2e-monitoring/tool-tests.sh]: Addressed in [d41e039](https://github.com/SolaceDev/solace-broker-mcp/pull/100/commits/d41e0390d8e2b348bac7e2bad5c34b253b55276a)

**Review by solace-aross — APPROVED** [review]
Both follow-ups look good. The tx threshold bump to 50 restores the regression signal, and keeping the per-call abort is fair given the failure history is rate values, not call errors. CI is green including the e2e-monitoring run at the new threshold.

One leftover doc mismatch, fine to fix in a follow-up or fold into your next PR: the "F4 — message rates" bullet in test/e2e-monitoring/README.md still says "Asserted against `≥ 40`"; the table and both scripts now say 50.
